Update type_traits.h.

LocutusOfBorg opened the pull request:
A regression was introduced in 2b80e29803d60f01701a67bc57ef06dacfe8af8b. In case of THRUST_DEVICE_SYSTEM_OMP build, the invoke_result_t is undefined due to missing #else branch. See e.g. https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1064730

LocutusOfBorg committed:
Update type_traits.h.

A regression was introduced in 2b80e29803d60f01701a67bc57ef06dacfe8af8b. In case of THRUST_DEVICE_SYSTEM_OMP build, the invoke_result_t is undefined due to missing #else branch.
See e.g. https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1064730

stanleytsang-amd reviewed (approved):
looks good to me, but also tagging @Naraenda to review

Naraenda commented:
I'm compiling the following snippet:
```cpp
// omp_test.cpp

#include <cstdlib>
#include <iostream>

#include <omp.h>

#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/sort.h>
#include <thrust/transform.h>


int main() {
#ifdef _OPENMP
    std::cout << _OPENMP << std::endl;
#endif

    int N = 160000000;
    thrust::host_vector<int> h_data(N);

    for (auto &num : h_data) {
        num = rand();
    }

    thrust::device_vector<int> d_data(h_data);
    thrust::sort(d_data.begin(), d_data.end());

    std::cout << d_data[N / 2 - 1] << "\n"
              << d_data[N / 2] << "\n"
              << d_data[N / 2 + 1] << std::endl;

    return 0;
}
```

Compiling with the following works and also prints the OpenMP version (but also works without this patch):

```sh
hipcc -o omp_test omp_test.cpp -fopenmp -Xopenmp-target=amdgcn-amd-amdhsa -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_OMP && ./omp_test`
```

Though dropping `-Xopenmp-target=` seems to fail (or perhaps I'm missing something!), but the patch seem to fix some of it:

```sh
hipcc -o omp_test omp_test.cpp -fopenmp -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_OMP && ./omp_test
```

I'm not sure if the intent is also to have rocThrust work with OpenMP targeting host, or just fix OpenMP offloading to the GPU.

CC: @stanleytsang-amd @LocutusOfBorg

Naraenda commented:
Just as I wrote that, I found out the cause of the rest of the build errors. `detail/nv_target.h` contains a similar structure as the one modified in the first commit of this PR. This causes various symbols to not be defined when using the OpenMP backend (`NV_IS_HOST`, `NV_IS_DEVICE`, `NV_IF_TARGET`, etc.). Simply reusing `target.h` from the HIP system should resolve the issue.

---

Adding this patch on top of this PR will fix this:
```patch
diff --git a/thrust/detail/nv_target.h b/thrust/detail/nv_target.h
index 1f93a09..131ca72 100644
--- a/thrust/detail/nv_target.h
+++ b/thrust/detail/nv_target.h
@@ -27,4 +27,8 @@
 #  include <thrust/system/hip/detail/nv/target.h>
 #elif THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
 #  include <nv/target>
+#else
+// When using a different backend, e.g. OpenMP, we still need to define the target.
+// We can simply reuse the HIP targets for this.
+#  include <thrust/system/hip/detail/nv/target.h>
 #endif
```

Please verify if this resolves the issue on your end too 😃

LocutusOfBorg commented:
I already fixed with my chunk, but I'm happy to also add your one! or feel free to add it if you have access to my PR code

Naraenda committed:
fix(detail/nv_target.h): include missing target macros when targetting a non-hip backend (i.e. openmp)

This change causes the HIP system target.h to be included when targetting a device system that is not HIP or CUDA.

Naraenda reviewed (approved):
Thanks! I've added it.

@stanleytsang-amd could you re-review?

LocutusOfBorg commented:
I like it!

Naraenda commented:
@stanleytsang-amd Is there anything else holding back the merge or are we good to go?

stanleytsang-amd commented:
Apologies.  CI looks good, I'll merge it in now.
